libbeat/cmd/instance/beat.go

```diff
@@ -783,6 +790,36 @@ func (b *Beat) templateLoadingCallback() (func(esClient *elasticsearch.Client) e
 	return callback, nil
 }
 
+// Build and return a callback to fetch the Elasticsearch cluster_uuid for monitoring
+func (b *Beat) clusterUUIDFetchingCalling() (elasticsearch.ConnectCallback, error) {
+	var response struct {
+		ClusterUUID string `json:"cluster_uuid"`
+	}
+
+	callback := func(esClient *elasticsearch.Client) error {
+		status, body, err := esClient.Request("GET", "/", "", nil, nil)
+		if err != nil {
+			return errw.Wrap(err, "error querying /")
+		}
+		if status > 299 {
+			return fmt.Errorf("Error querying /. Status: %d. Response body: %s", status, body)
+		}
+		err = json.Unmarshal(body, &response)
+		if err != nil {
+			return fmt.Errorf("Error unmarshaling json when querying /. Body: %s", body)
+		}
+
+		stateRegistry := monitoring.GetNamespace("state").GetRegistry()
+		outputsRegistry := stateRegistry.NewRegistry("outputs")
+		elasticsearchRegistry := outputsRegistry.NewRegistry("elasticsearch")
+		monitoring.NewString(elasticsearchRegistry, "cluster_uuid").Set(response.ClusterUUID)
+
+		return nil
+	}
+
+	return callback, nil
```

Comment:
The callback will be run on every reconnect. The first connect only occurs if the beat actually has an event to publish. If there is no event or the output is not available on start we can not fetch the cluster id at all. Will this be an issue in the UI?

If possible I'd prefer not to include the ES cluster ID.

Reply:
> ... or output is not available on start

Can you clarify what this means or how this scenario can occur?

> If possible I'd prefer not to include the ES cluster ID.

Include it in the monitoring events/docs? Or somewhere else?

Reply:
>> ... or output is not available on start
> Can you clarify what this means or how this scenario can occur?

When beats startup, but is idle, then we do not attempt to connect to Elasticsearch. That is, the callback checking for the cluster ID is never run. The callback will only applied if we have events to publish AND we sucessfully managed to connect and ping the remote ES host. If connection fails and event is eventually dropped (in case of metricbeat/packetbeat), then the callback will still not be executed.

Due to this we have no cluster ID, yet metrics are/must be published to the monitoring cluster all the time.

>> If possible I'd prefer not to include the ES cluster ID.
> Include it in the monitoring events/docs? Or somewhere else?

Not send the ES cluster ID at all. There is no guarantee if/when you have the cluster ID available. E.g. if the beat drops all events because the host setting is wrong, then we still want to get the monitoring metrics.
If the beat is configured to ship via Logstash/Kafka/Redis, then you won't have a cluster ID.

Reply:
I understand better now, thanks for the explanation.

> ...
> Due to this we have no cluster ID, yet metrics are/must be published to the monitoring cluster all the time.

Technically it should be okay to send monitoring docs without a `cluster_uuid`. The Monitoring UI will be able to handle that once https://github.com/elastic/kibana/issues/27595 is resolved.

> ...
> If the beat is configured to ship via Logstash/Kafka/Redis, then you won't have a cluster ID.

Good point. In this case we could ask users to configure `monitoring.elasticsearch.*`. 

--

Bottomline is that the `cluster_uuid` corresponds to the ES cluster that the beat is sending its data to via `output.elasticsearch.*`. If this output is not configured or we are not able to connect to that cluster yet or ever, there's no way we can figure out its `cluster_uuid`. So there's no way we can, or even should, try to associate monitoring data with this cluster.

Reply:
right.

In the end it’s a correlation problem. It would be easier for us if we would send the/one hostname and the monitoring endpoint will do the lookup in order to fill the cluster id.

> Good point. In this case we could ask users to configure monitoring.elasticsearch.*

makes sense. But this Elasticsearch might not be reachable. For this case it we could also introduce a `montoring.elasticsearch.cluster_id` setting.

Reply:
We had this exact discussion amongst the stack monitoring team. In fact, the original plan was to include exactly a setting like you mentioned. The concern was that it puts adds to the user's burden and, after discussion, we decided a better solution was to make it so that the Monitoring UI could handle monitoring data with no `cluster_uuid` in it. So we are going down that path now: https://github.com/elastic/kibana/pull/28278.